### PR TITLE
Fix panic on writing to closed channel

### PR DESCRIPTION
**Pull request by anujva**
The original issue was opened against the moby docker project and reported here: https://github.com/moby/moby/issues/42316

The docker daemon RingLogger implementation is written in such a way
that it can call Close() on the underlying driver, without stopping to
write to it. This can cause panics on fluent-golang-logger which has no
checks on whether the channel was closed by the daemon. The contract on
the logger.Logger interface in docker doesn't specify that it will make
sure that the call to Close() and Log() will not compete so other
logging driver implementations have similar semantics as this commit to
account for that.

The relevant stacktrace on panic:
```
github.com/docker/docker/vendor/github.com/fluent/fluent-logger-golang/fluent.(*Fluent).appendBuffer(0xc0043a11e0, 0xc0
084da1e0, 0xc0084da180, 0xc0084da1b0)
        /builddir/build/BUILD/docker-19.03.6ce-4.58.amzn1/src/github.com/docker/docker/vendor/github.com/fluent/fluent-
logger-golang/fluent/fluent.go:306 +0x4c
github.com/docker/docker/vendor/github.com/fluent/fluent-logger-golang/fluent.(*Fluent).postRawData(0xc0043a11e0, 0xc00
84da1e0, 0x1d, 0x1668656)
        /builddir/build/BUILD/docker-19.03.6ce-4.58.amzn1/src/github.com/docker/docker/vendor/github.com/fluent/fluent-
logger-golang/fluent/fluent.go:220 +0x3d
github.com/docker/docker/vendor/github.com/fluent/fluent-logger-golang/fluent.(*Fluent).EncodeAndPostData(0xc0043a11e0,
 0xc00ab8fa80, 0x1d, 0x1668656, 0xed814becb, 0x0, 0x560000475460, 0xc0084da180, 0x98, 0xc003067a68)
        /builddir/build/BUILD/docker-19.03.6ce-4.58.amzn1/src/github.com/docker/docker/vendor/github.com/fluent/fluent-
logger-golang/fluent/fluent.go:210 +0x159
github.com/docker/docker/vendor/github.com/fluent/fluent-logger-golang/fluent.(*Fluent).PostWithTime(0xc0043a11e0, 0xc0
0ab8fa80, 0x1d, 0x1668656, 0xed814becb, 0x0, 0x56000047ab60, 0xc0084da150, 0x0, 0xc001770a80)
        /builddir/build/BUILD/docker-19.03.6ce-4.58.amzn1/src/github.com/docker/docker/vendor/github.com/fluent/fluent-
logger-golang/fluent/fluent.go:201 +0x753
github.com/docker/docker/daemon/logger/fluentd.(*fluentd).Log(0xc008573cc0, 0xc0001e4c40, 0x0, 0x0)
        /builddir/build/BUILD/docker-19.03.6ce-4.58.amzn1/src/github.com/docker/docker/daemon/logger/fluentd/fluentd.go
:177 +0x3cb
github.com/docker/docker/daemon/logger.(*RingLogger).run(0xc00d8e2a50)
        /builddir/build/BUILD/docker-19.03.6ce-4.58.amzn1/src/github.com/docker/docker/daemon/logger/ring.go:119 +0x7d
created by github.com/docker/docker/daemon/logger.newRingLogger
        /builddir/build/BUILD/docker-19.03.6ce-4.58.amzn1/src/github.com/docker/docker/daemon/logger/ring.go:43 +0x176
```

Signed-off-by: Anuj Varma <anujvarma@thumbtack.com>

### CODE REVIEW

**Comment by anujva** [comment]
If the approach is acceptable, we can update the unit tests for this.

File: fluent/fluent.go [hunk 1]
```diff
@@ -338,6 +344,11 @@ func (f *Fluent) Close() (err error) {
 
 // appendBuffer appends data to buffer with lock.
 func (f *Fluent) appendBuffer(msg *msgToSend) error {
+	f.pendingMutex.RLock()
+	defer f.pendingMutex.RUnlock()
+	if f.chanClosed {
+		return fmt.Errorf("fluent#appendBuffer: Channel was already closed by docker daemon")
```
Comment (toraora) [inline fluent/fluent.go:350]: error message should be more generic, like "Logger is closed" or something

**Comment by tagomoris** [comment]
This change looks good. @toraora 's comment about the error message looks reasonable to me too.

File: fluent/fluent.go (hunk 1 above)
Comment (anujva) [reply fluent/fluent.go]: Updated.

**Comment by tagomoris** [comment]
@anujva I missed mentioning tests. Could you add a test case about this fix?

**Comment by anujva** [comment]
> @anujva I missed mentioning tests. Could you add a test case about this fix?

Yep.. I will update it. Thanks!

**Comment by anujva** [comment]
> > @anujva I missed mentioning tests. Could you add a test case about this fix?
> 
> Yep.. I will update it. Thanks!

Hi @tagomoris 
Let me know if the test cases makes sense or if I should tweak them.

**Comment by tagomoris** [comment]
@anujva your test case looks to simulate the actual situation well but also seems to cause random test failures due to goroutine scheduling (probably, very rarely).
I thought it's enough to have a test case to just `f.Close` the logger and then call `f.EncodeAndPostData`. Isn't it good enough in your idea?

**Comment by anujva** [comment]
> @anujva your test case looks to simulate the actual situation well but also seems to cause random test failures due to goroutine scheduling (probably, very rarely).
> I thought it's enough to have a test case to just `f.Close` the logger and then call `f.EncodeAndPostData`. Isn't it good enough in your idea?

I checked the test.. the tests were failing was Logf is not concurrency safe it seems.. I removed the logging of the count parameter from the test, and then ran the test 10000 times, using this command 
`go test -v -run TestNoPanicOnAsyncClose -count 10000` . It does not fail now. 

If it still fails, I will gladly change it to what you suggested. Please let me know. Thanks.

**Comment by tagomoris** [comment]
My idea was to make the test case as simple as possible. For example, we can check the fix is working well with the code like:
```go
t.Run("this should not panic", func(t *testing.T) {
	// ...
	f, err := newWithDialer(testcase.config, d)
	f.Close()
	tag := "myapp.access"
	var data = map[string]string{
		"foo":  "bar",
		"hoge": "hoge",
	}
	error := f.Post(tag, data)
	assert.Equal(t, fmt.Errorf("fluent#appendBuffer: Logger already closed"), e)
})
```

The test case in this change should work well on our laptops but may work incorrectly sometimes when CI environment had just a single core and it doesn't switch context well. I have a motivation to keep test cases simple for those things.

**Comment by anujva** [comment]
I have updated the test case to just run the two scenarios in a single go
routine. Please let me know if that looks good.

**Comment by tagomoris** [comment]
@anujva LGTM! Thanks for updating the test case.

**Comment by tagomoris** [comment]
Merged, and tagged `v1.6.0`

**Pull request merged**